nougat.1
.\" Automatically generated by Pandoc 2.1.2
.\"
.TH "nougat" "1" "" "Version 1.0" "Nougat Man Page"
.hy
.SH NAME
.PP
nougat \- A beautiful screenshot wrapper
.SH SYNOPSIS
.PP
\f[I]nougat\f[] [ \f[B]\-h\f[] | \f[B]\-f\f[] | \f[B]\-m\f[] | \f[B]\-c\f[] |
\f[B]\-t\f[] | \f[B]\-s\f[] | \f[B]\-i\f[] | \f[B]\-b\f[] \f[I]backend\f[] |
\f[B]\-p\f[] ] [\f[I]DIRECTORY\f[]]
.SH DESCRIPTION
.PP
\f[B]Nougat\f[] is a screenshot wrapper that organizes screenshots by
date and time (dates are subdirectories, times are filenames).
It has backend support for maim, scrot, and imagemagick, as well as a
few convenient features for users.
.SH OPTIONS
.PP
\f[B]\-h\f[]
.PD 0
.P
.PD
Display the help and exit.
.PP
\f[B]\-f\f[]
.PD 0
.P
.PD
Takes a fullscreen screenshot.
.PP
\f[B]\-m\f[]
.PD 0
.P
.PD
Takes a fullscreen screenshot of the currently focused display.

This requires xdotool and is incompatible with the scrot backend.
.PP
\f[B]\-c\f[]
.PD 0
.P
.PD
Copies to clipboard (Requires xclip dependency)
.PP
\f[B]\-t\f[]
.PD 0
.P
.PD
Places the screenshot in /tmp rather than the set directory.
.PP
\f[B]\-i\f[]
.PD 0
.P
.PD
Outputs the taken image to stdout. This is useful for scripts
that accept an image as input on stdin. (Implies \-s)
.PP
.PP
\f[B]\-s\f[]
.PD 0
.P
.PD
Does not output the file's path.
.PP
\f[B]\-b\f[]
.PD 0
.P
.PD
Selects what backend to use.
Nougat will detect avaliable backends if \-b is not specified.
Nougat will prefer maim over all others.
.PP
\f[B]\-p\f[]
.PD 0
.P
.PD
Cleans symlinks that point to nowhere in the all subdirectory.
Does not take screenshot.
.PP
\f[B]DIRECTORY\f[]
.PD 0
.P
.PD
Puts current screenshot in specified directory.
.SH EXAMPLES
.PP
\f[B]nougat \-fc\f[] \f[I]~/pictures\f[]
.PD 0
.P
.PD
\ \ \ \ This particular command takes a fullscreen picture without a
cursor and moves it to clipboard as well as saving it directly to
~/pictures/
.PP
\f[B]nougat \-cts\f[]
.PD 0
.P
.PD
\ \ \ \ This command here will allow you to do selection, and take the
picture without cursor and silently without putting it to any directory
but putting it directly into clipboard.
.PP
\f[B]nougat \-c\f[]
.PD 0
.P
.PD
\ \ \ \ This will take a picture of your screenshot in selection and
without cursor but put the picture into whatever your
NOUGAT_SCREENSHOT_DIRECTORY is in the symlinked to all and
subcatagorized by date folders.
.PP
\f[B]nougat \-fcb imagemagisk\f[] \f[I]~/pictures\f[]
.PD 0
.P
.PD
\ \ \ \ This command will take a full screen picture with cursor and
saves it to ~/pictures as well as clipboard, however it requests
imagemagisk to be the backend instead of whatever is originally used.
.PP
\f[B]nougat \-fi | curl -sF'file=@-' https://0x0.st\f[]
.PD 0
.P
.PD
\ \ \ \ This takes a fullscreen picture and pipes
the image to curl where it is uploaded to 0x0.st (requires curl installed).
.PP
.SH ENVIRONMENT
.PP
\f[B]NOUGAT_CONFIG\f[]
.PD 0
.P
.PD
\ \ \ \ This is your configuration, All environment variables can be
modified from this file.
It is found in your configuration folder.
\f[B]EX:~/.config/nougat\f[]
.SH TROUBLESHOOTING
.PP
\f[B]I've just installed Nougat, and set it to a keybinding, but it
doesn't work with the keybinding.\f[]
.PD 0
.P
.PD
\ \ \ \ First ensure that you meet the dependencies of at least one
backend.
If you've exported the screenshot directory, you will be required to
logout and log back in.
.PP
\f[B]Nougat always outputs the path of a screenshot.\f[]
.PD 0
.P
.PD
\ \ \ \ This is done to make it easier to integrate Nougat into other
applications (for example an image uploader).
Pass \f[B]\-s\f[]
.PP
\f[B]My name is Ted\f[]
.PD 0
.P
.PD
\ \ \ \ Get the developer and general Linux users of varying experience
levels together for an hour in a Discord voice chat, one explaining a
handful of commands to install it, the other two making a pacman package
for you.
Then, one puts the package on the AUR, \ \ \ \ and install from there.
.SH AUTHOR
.PP
Nougat was created by Sweets <https://github.com/Sweets>
.PD 0
.P
.PD
Man Page was created by Good Friend Snow.
.PD 0
.P
.PD
Fucking Ted.


nougat.sh
#!/usr/bin/env bash

# Nougat version 2

# All features from nougat 1
# ~/.config/nougat

saveourship() {
cat << EOF
Nougat - screenshot wrapper created to help organize screenshots
 -h - Saves our ship.
 -i - Output image to stdout. This implies -s (silent).
 -s - Silent. By default, nougat will output the path to the file to STDOUT.
              This is to make it easier to implement into other file uploaders.
 -t - Places screenshot into /tmp
      (useful if you only need a quick screenshot to send to a friend)
 -f - Takes a full screen screenshot (default is select area)
 -m - Takes a full screen screenshot of the currently focused monitor.
      \`xdotool\` is required (\`nougat\` fallsback to a regular fullscreen screenshot without it).

 -c - Puts the screenshot into your clipboard
 -b - Select backend to use
              Supported backends are \`maim', \`scrot' (without -m), and \`imagemagick'.
              nougat will detect available backends if -b
              is not specified. nougat prefers maim to scrot and imagemagick.
 -p - Cleans the link directory of Nougat based on the linking policy.
              Particularly useful as it cleans any links that no
              longer point to a screenshot (i.e. deleted screeenshot).
EOF
}

temporary=false
clean=false
silent=false
stdout=false
fullscreen=false
focused_monitor=false
copytoclipboard=false
backend=""

backends=('maim' 'scrot' 'imagemagick')

### BACKENDS

maimbackend() {
    require maim

    maimopts=''

    [[ $fullscreen == false ]] && maimopts=-s
    [[ $focused_monitor == true ]] && maimopts=--geometry=$(getcurrentmonitor)
    [[ $maimopts == "--geometry=" ]] && maimopts=''
    maimopts="$maimopts --hidecursor"

    maim $maimopts /tmp/nougat_temp.png
}

scrotbackend() {
    require scrot

    scrotopts=''

    [[ $fullscreen == false ]] && scrotopts=-s

    scrot $scrotopts /tmp/nougat_temp.png
}

imagemagickbackend() {
    require import

    importopts=''

    [[ $fullscreen == false && $focused_monitor == false ]] && {
        require slop

        slop=$(slop -qof '%wx%h+%x+%y')

        [[ -n $slop ]] && importopts="-crop $slop"
    } || {
      [[ $focused_monitor == true ]] && importopts="-crop $(getcurrentmonitor)"
      [[ $importopts == "-crop " ]] && importopts=''
    }

    import -window root $importopts /tmp/nougat_temp.png
}

### END BACKENDS

nobackend() {
    if [[ -z $backend ]]; then
        return 0
    else
        return 1
    fi
}

testfor() {
    command -v "$1" &> /dev/null

    return "$?"
}

require() {
    command -v "$1" &> /dev/null || {
        echo "$1 is not installed and is required"
        exit 1
    }
}

getconfigdir() {
    CONFIG_DIR="$XDG_CONFIG_HOME"

    [[ ! -d $CONFIG_DIR ]] && CONFIG_DIR="$HOME/.config"

    echo "$CONFIG_DIR"
}

# Print's the current screens geometry
# Modified from:
# https://superuser.com/a/1238384
getcurrentmonitor() {
  # We require xdotool to get the mouse co-ordinates (used to tell which monitor we're on)
  # Without xdotool we silently fallback to a regular fullscreen screenshot
  testfor xdotool || return 1
  testfor xrandr  || return 1

  xrandr="$(xrandr --nograb)"
  [[ -z "$xrandr" ]] && return 1

  OFFSET_RE="\+([-0-9]+)\+([-0-9]+)"

  # Get the window position
  eval "$(xdotool getmouselocation --shell)"

  # Loop through each screen and compare the offset with the window
  # coordinates.
  monitor_index=0
  while read -r name width height xoff yoff
  do
      if [[ "${X}" -ge "$xoff" && \
            "${Y}" -ge "$yoff" && \
            "${X}" -lt "$(($xoff+$width))" && \
            "${Y}" -lt "$(($yoff+$height))" ]]; then
          monitor=$name
          break
      fi
      ((monitor_index++))
  done < <(grep -w connected <<< "$xrandr" |
      sed -r "s/^([^ ]*).*\b([-0-9]+)x([-0-9]+)$OFFSET_RE.*$/\1 \2 \3 \4 \5/" |
      sort -nk4,5)

  # If we found a monitor, get its geometry.
  if [[ "$monitor" ]]
  then
      geometry="$(grep -E "^$monitor\s" <<< "$xrandr")"
      geometry="${geometry/$monitor/}"
      geometry="${geometry/connected/}"
      geometry="${geometry/primary/}"
      awk '{print $1}' <<< "$geometry"
      return $?
  fi

  return 1
}

getcanonicals() {
    read -r year month day hour minute second <<< "$(date +'%Y %B %d %H %M %S')"

    suffix=''
    [[ $fullscreen == true ]] && suffix=_full

    . "$(getconfigdir)/nougat"

    ORG_FULLPATH="$NOUGAT_SCREENSHOT_DIRECTORY/$NOUGAT_ORGANIZATION_POLICY"

    [[ -n $NOUGAT_LINKING_POLICY ]] &&
      LINK_FULLPATH="$NOUGAT_SCREENSHOT_DIRECTORY/$NOUGAT_LINKING_POLICY" ||
      LINK_FULLPATH=""

    echo "$(dirname "$ORG_FULLPATH")" \
        "$(basename "$ORG_FULLPATH")" \
        "$([[ -n $LINK_FULLPATH ]] && dirname  "$LINK_FULLPATH")" \
        "$([[ -n $LINK_FULLPATH ]] && basename "$LINK_FULLPATH")"
}

init() {
    CONFIG_DIR=$(getconfigdir)

    [[ ! -f $CONFIG_DIR/nougat ]] && {
        mkdir -p "$CONFIG_DIR"

        if [[ -n $NOUGAT_SCREENSHOT_DIRECTORY ]]; then
            # Support for V1 configurations
            echo 'NOUGAT_SCREENSHOT_DIRECTORY="'"$NOUGAT_SCREENSHOT_DIRECTORY"'"' > "$CONFIG_DIR/nougat"
        else
            echo 'NOUGAT_SCREENSHOT_DIRECTORY="$HOME/Screenshots"' > "$CONFIG_DIR/nougat"
        fi

        cat >> "$CONFIG_DIR/nougat" << EOF
NOUGAT_ORGANIZATION_POLICY="\${year}/\${month}/\${day}/\${hour}:\${minute}:\${second}\${suffix}"
NOUGAT_LINKING_POLICY="All/\${year}-\${month}-\${day}.\${hour}:\${minute}:\${second}\${suffix}"
EOF
    }

    while getopts 'hstfmcpui b:S:' option; do
        case $option in
            h)
                saveourship
                exit 0
                ;;
            b) setbackend $OPTARG;;
            # Hide cursor
            p) clean=true;;
            s) silent=true;;
            t) temp=true;;
            c) copytoclipboard=true;;
            f) fullscreen=true;;
            m)
               fullscreen=true
               focused_monitor=true
              ;;
            i)
               stdout=true
               silent=true
               ;;
        esac
    done

    local exitcode; exitcode="$?"

    nobackend &&
        testfor maim && backend=maim && return "$exitcode"

    nobackend &&
        testfor scrot && backend=scrot && return "$exitcode"

    nobackend &&
        testfor import && backend=imagemagick && return "$exitcode"
}

setbackend() {
    supported=false

    for (( index=0; index<${#backends}; index++ )); do
        [[ ${backends[$index]} == "$1" ]] && {
            supported=true
            break
        }
    done

    [[ $supported == false ]] && {
        echo "Unsupported backend $1"
        exit 1
    }

    cmd="$1"

    [[ $cmd == imagemagick ]] && cmd=import

    testfor $cmd && backend="$1"
}

runbackend() {
    case $backend in
        maim) maimbackend;;
        scrot) scrotbackend;;
        imagemagick) imagemagickbackend;;
        *)
            echo 'No supported backend found'
            exit 1
            ;;
    esac

    [[ ! -f /tmp/nougat_temp.png ]] && exit 0
}

stdout() {
  if [[ ! -t 1 ]]; then
    cat "$1"
  else
    cat >&2 <<EOF
Refusing to output stdout to a terminal.

Did you mean to pipe me to another script?

$1
EOF
    exit 1
  fi
}

organize() {
    read -r fullpath filename linkpath linkname <<< "$(getcanonicals)"

    [[ $copytoclipboard == true ]] && {
        require xclip
        xclip -selection clipboard -t image/png /tmp/nougat_temp.png
    }

    [[ $temp == true ]] && {
        [[ $silent == false ]] && echo "/tmp/$linkname.png"

        mv /tmp/nougat_temp.png "/tmp/$linkname.png"

        [[ $stdout == true ]] && stdout "/tmp/$linkname.png"

        exit 0
    }

    mkdir -p "$fullpath"
    [[ -n $linkpath ]] && mkdir -p "$linkpath"

    mv /tmp/nougat_temp.png "$fullpath/$filename.png"
    [[ -n $linkpath ]] && ln -s "$fullpath/$filename.png" "$linkpath/$linkname.png"

    [[ $silent == false ]] && echo "$fullpath/$filename.png"

    [[ $stdout == true ]] && stdout "$fullpath/$filename.png"

    exit 0
}

clean() {
    . "$HOME/.config/nougat"

    linkdir=$(dirname "$NOUGAT_SCREENSHOT_DIRECTORY/$NOUGAT_LINKING_POLICY")
    [[ $silent == false ]] && echo "$linkdir"

    [[ ! -d $linkdir || -z $(ls -1 "$linkdir") ]] && return 0

    for file in "$linkdir/"*; do
        link=$(readlink -f "$file")

        if [[ ! -f $link ]]; then rm "$file"; fi
    done
}

screenshot() {
    runbackend
    organize
}

init "$@"

[[ $clean == true ]] && {
    clean
    x="$?"

    [[ $# -eq 1 ]] && {
        [[ $1 == -p || $1 == -ps || $1 == -sp ]] && exit "$x"
    }

    [[ $# -eq 2 ]] && {
        [[ $1 == -p || $1 == -s ]] && {
            [[ $2 == -p || $2 == -s ]] && exit "$x"
        }
    }
}

screenshot
